Added MacroPolo addon

thatdudegrim opened the pull request:
MacroPolo is a simple addon for macro navigation enabling the use of /macro back to return to the previous book and set.

This allows easy sharing between overlapping macro lines.

FFXIMacroEditor committed:
Create MacroPolo.lua

FFXIMacroEditor committed:
Added MacroPolo to addons.xml

FFXIMacroEditor committed:
Update MacroPolo.lua

Added support for //mp back as an alias to /macro back.

ghost commented:
Is this addon permissible? I know I hijack a unique syntax involving the /macro command but I wanted to make it easy for people to use it in macros without the console command as that confuses a lot of people.

I searched for a little bit but it seemed impossible to derive the the current macro position without the user specifying it.

z16 commented:
It's permissible, I just didn't have that much time to review it yet.

One thing to note is that you don't need the `split` function, as there is a `string.split` function:

``` lua
print('one two three':split(' ')) -- Will print "[one, two, three]"
```

I would advocate to not use that syntax, instead go back to `//macro`, but that's really up to you. I'll give a more detailed review soon, maybe tomorrow.

FFXIMacroEditor committed:
Update MacroPolo.lua

Migrated to string:split rather than custom function to split.

ghost commented:
Thanks,

Updated to use the built in split function.

Yeah, I mean I can use outgoing text to capture anything, and of course it _is_ running briefly at every occurrence of outgoing text, every **/p hi**, etc. Again, the only motivation for using this method was to let people easily use it in macros rather than try to find out why **//macropolo...** doesn't directly work in macros (instead, having to use say, **//console macropolo...**)

Since it is running at every outgoing text (every /p, every /ws, etc), would it benefit to use a simpler if surrounding the regex match like..

`if original:sub(1,6) == '/macro' then`
`if windower.regex.match...`

z16 commented:
Yes, it would be more efficient, though I'd use this:

``` lua
windower.register_event('outgoing text', function(original)
    if not original:startswith('/macro') then
        return
    end

    local current_book, current_set, target_book, target_set = original:split(' ', 2)[2]:match('(%d+)-(%d+) (%d+)-(%d+)')
    if not target_set then -- Failed to parse
        error('Invalid command')
    end

    -- Logic here
end)
```

However, I'm not convinced of the argument for `/macro` vs `//macro`. It has been a very well-known Windower shortcoming for many years, and I don't think anyone who would use an addon such as this does not know about it. And even if there are such people, this could be explained in a _ReadMe_ file.

Which brings me to your help text, imo it's too long and should really only be a one-liner about command usage. Explaining what it does is what a _ReadMe_ file is for.

I'll comment on it again at some point this weekend, sorry it's taking me so long.

FFXIMacroEditor committed:
Update MacroPolo.lua

Moved over to the /macropolo command, rather than /macro.

FFXIMacroEditor committed:
Create README.md

ghost commented:
Thanks. I understand you're a busy individual.

I never minded using **//macro** over **/macro**. My thoughts were that //macro won't work when used directly inside a macro. **/console ...** will, but I'm trying to make a simple syntax available. Over my years in FFXI, people have asked me several times "Why won't my macro execute this windower command?".

I tried to use **//macro**, with the outgoing text event and it would not fire. I don't run many addons and nothing unusual so perhaps something pretty typical is overriding it.

At one time, test code literally was

```
windower.register_event('outgoing text', function(original)
    print(original)
end)
```

Since I couldn't get it to work using **//macro** in macros, I migrated everything to support **/macropolo** (via outgoing text) and **//macropolo** & **/console macropolo ...** (via addon command). This way it will work for novices and people familiar with using windower commands in macros. It no longer employs /macro in any way.

I actually expect more people that use it to be new/returning, and working on macros for macro-intense jobs like Scholar, rather than people very comfortable with their current macro layout.

I also moved most of help text to a README.md as you suggested.

Thanks also for for the condensing of my regex using split().

Edited for typo.

ghost commented:
Also, I removed mp as an addon command. Since this isn't intended to be typed out each time, the length of the command makes little difference. I removed it mainly because it's too easy that people might use mp as an alias for something, since it's a game term.

FFXIMacroEditor committed:
Update addons.xml

Updated description for MacroPolo addon in addons.xml.

ghost commented:
Hi, any updates?

z16 commented:
Sorry, I'll review this in the coming days.

FFXIMacroEditor committed:
Create trustworthy.lua

Trustworthy is a simple addon for searching your trusts since the game will continue to offer you ciphers for trusts you have.

FFXIMacroEditor committed:
Update addons.xml

Added Trustworthy addon.

FFXIMacroEditor committed:
Update trustworthy.lua

FFXIMacroEditor committed:
Update trustworthy.lua

z16 commented:
Ouch, sorry, this totally passed me by. I'll review it now, but I'll note one thing: There are suddenly two addons here. Could you please remove _TrustWorthy_ from this and submit a separate pull request for it? I'm gonna focus on reviewing _MacroPolo_ here. That makes it easier to handle them separately.

z16 commented on addons/MacroPolo/MacroPolo.lua:
```diff
@@ -0,0 +1,72 @@
+        log("Please see README.md for more details.")
+    elseif args[1] == "back" then
+        if #mHistory then
+            windower.send_command("input /macro book "..mHistory[#mHistory][1].."; input /macro set "..mHistory[#mHistory][2])
```
Should use `windower.chat.input('/something')` instead of `windower.send_command('input /something')`. That accesses FFXI's input box directly and doesn't go through our console handler. A bit more efficient and less error prone. Should also split the two to separate function calls:

``` lua
            windower.chat.input('/macro book ' .. mHistory[#mHistory][1])
            windower.chat.input('/macro set ' .. mHistory[#mHistory][2])
```

z16 commented on addons/MacroPolo/MacroPolo.lua:
```diff
@@ -0,0 +1,72 @@
+        local current_book, current_set, target_book, target_set = argstring:match('(%d+)-(%d+) (%d+)-(%d+)$')
+        
+        if target_set then
+            windower.send_command("input /macro book "..target_book.."; input /macro set "..target_set)
```
See above.

z16 commented on addons/MacroPolo/MacroPolo.lua:
```diff
@@ -0,0 +1,72 @@
+        
+        if target_set then
+            windower.send_command("input /macro book "..target_book.."; input /macro set "..target_set)
+            mHistory[#mHistory+1] = {current_book,current_set}
```
Just to make the code more readable I'd suggest the following things, but they're really up to you:
1. I'd use our lists library: `mHistory = L{}`. Then you can add an element by doing: `mHistory:add({book, set})`. You don't have to work with `#` to get the length all the time. This helps above as well, you can just do `mHistory:last()` to get the last element (or `mHistory:last(2)` to get the second to last, and so on).
2. I'd use named table keys for the book/set record: `{book = current_book, set = current_set }`. That way you can do `mHistory:last().book` and `mHistory:last().set`, instead of using `1` and `2`, which does not make it obvious what it's supposed to do.

z16 commented on addons/MacroPolo/README.md:
```diff
@@ -0,0 +1,11 @@
+
+You may then use either of these commands in a macro to return to the previous book and set.
+/macropolo back
+/console macropolo back
```
Nothing to critique here, but I'd beef up the file with some markdown. That's why it's called `.md` after all :)

z16 commented on addons/MacroPolo/MacroPolo.lua:
```diff
@@ -0,0 +1,72 @@
+    end
+end
+
+windower.register_event('outgoing text', function(original)
```
As I mentioned before, I'm not a fan of reacting to both outgoing text and addon commands separately, but again, this is up to you. I'd pick one (addon command) and only use that, and trust the users to be smart enough to figure it out. If you insist on outgoing text, I'd make it react to just `/macro`, so it looks like it's extending the in-game macro command. Right now it's separate from the in-game command, yet separate from standard Windower commands as well, in a strange middle place.

z16 commented on addons/MacroPolo/MacroPolo.lua:
```diff
@@ -0,0 +1,72 @@
+
+windower.register_event('outgoing text', function(original)
+    if original:startswith('/macropolo ') then
+        Commander(original:sub(tonumber(original:find(" "))))
```
`string.find` already returns a number, the `tonumber` call is unnecessary.
